### PR TITLE
Add a few options to prevent adding / removing / moving of vertices and add Keyboard functions

### PR DESCRIPTION
**Pull request by Falke-Design**
New Edit-Mode options:
```
addVertexOn: 'click',
removeVertexOn: 'contextmenu',
removeVertexValidation: undefined, 
addVertexValidation: undefined,
moveVertexValidation: undefined
```

And add Keyboard functions `map.pm.Keyboard`:
> We implemented a built-in keyboard listener to make one central place where keyboard events can be accessed (without adding self a listener).
```
  getLastKeyEvent(type = 'current'),
  isShiftKeyPressed(),
  isAltKeyPressed(),
  isCtrlKeyPressed(),
  isMetaKeyPressed(),
  getPressedKey(),
```

Now we can add a check to remove / add / move vertices for example with ctrl hold:
```
const check = (obj)=>{
  return map.pm.Keyboard.isCtrlKeyPressed();
};
map.pm.setGlobalOptions({removeVertexValidation: check, addVertexValidation: check, moveVertexValidation: check})
```

Feature: #817

### CODE REVIEW

File: src/js/Mixins/Keyboard.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  _initKeyListener(map) {
+    this.map = map;
+    L.DomEvent.on(document, 'keydown keyup', this._onKeyListener, this);
+  },
+  _onKeyListener(e) {
+    let focusOn = 'document';
+
+    // TODO: .contains only supported since IE9, problem? Should we define, which Versions we are Support?
+    // with focusOn the user can add a check if the key was pressed while the user interacts with the map
+    if (this.map.getContainer().contains(e.target)){
```
Comment (Falke-Design) [inline src/js/Mixins/Keyboard.js:12]: We will not support IE9

File: src/js/Mixins/Keyboard.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    const data = {event: e, eventType: e.type, focusOn};
+    this._lastEvents[e.type] = data;
+    this._lastEvents.current = data;
+
+    L.PM.Utils._fireEvent(this.map, 'pm:keyevent',data);
+  },
+  getLastKeyEvent(type = 'current') {
+    return this._lastEvents[type];
+  },
+  // TODO: isXYZPressed or getXYZPressed?
```
Comment (Falke-Design) [inline src/js/Mixins/Keyboard.js:25]: we use `is...`

**Review by Falke-Design — COMMENTED** [review]
Use in snapping the KeyboardMixin

File: src/js/Edit/L.PM.Edit.js [hunk 3]
```diff
@@ -35,7 +40,48 @@ const Edit = L.Class.extend({
...
+      return false;
+    }
+    return true;
+  },
+  _vertexValidationDragEnd(marker){
+    if(marker._vertexResetLatLng){
+      marker._vertexResetLatLng = null;
+      return false;
+    }
+    return true;
```
Comment (codeofsumit) [inline src/js/Edit/L.PM.Edit.js:82]: these validations don't seem to validate anything 🤔 
Also: where are the users validation functions being called?

**Review by codeofsumit — CHANGES_REQUESTED** [review]

File: leaflet-geoman.d.ts [hunk 4]
```diff
@@ -703,6 +742,8 @@ declare module 'leaflet' {
             allowSelfIntersection?: boolean;
         }
 
+        type VertexValidationHandler = (e: { layer: L.Layer, marker: L.Marker, event: any }) => boolean;
```
Comment (Falke-Design) [inline leaflet-geoman.d.ts:745]: is this correct? It should be a function with the arguments `layer`, `marker`, `event` and returns a boolean value.

File: leaflet-geoman.d.ts [hunk 5]
```diff
@@ -716,12 +757,39 @@ declare module 'leaflet' {
...
+            addVertexOn?:
+                | 'click'
+                | 'dblclick'
+                | 'mousedown'
+                | 'mouseover'
+                | 'mouseout'
+                | 'contextmenu';
+
+            /** A function for validation if a vertex (of a Line / Polygon) is allowed to add. It passes a object with `[layer, marker, event}`. For example to check if the layer has a certain property or if the `Ctrl` key is pressed. (default:undefined). */
+            addVertexValidation?: undefined | VertexValidationHandler;
```
Comment (Falke-Design) [inline leaflet-geoman.d.ts:776]: is this working? undefined or the function

**Review by Falke-Design — COMMENTED** [review]
@ryan-morris can you please look over this TypeScript implementation

File: leaflet-geoman.d.ts (hunk 5 above)
Comment (ryan-morris) [reply leaflet-geoman.d.ts:776]: These 2 scenarios are valid based on these changes:

```
const b: L.PM.DrawModeOptions = {
    addVertexValidation: undefined,
};

const c: L.PM.DrawModeOptions = {
    addVertexValidation: e => { return false; },
};
```

Comment (Falke-Design) [reply leaflet-geoman.d.ts:776]: Ok thanks

**Review by codeofsumit — CHANGES_REQUESTED** [review]

**Pull request merged**